From 59065ccc7781f762ef1ab1fbbb0b5e9769a69c7c Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Mon, 14 Sep 2020 14:44:51 +0300
Subject: [PATCH] [SW-17195] preboot: lkd: add fuse version

Add a way for LKD to read the device ID stored in eFuse

Change-Id: I0188495604dad90cf53e36478cb972fbbb39ae03
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/preboot/preboot.c b/preboot/preboot.c
index 6c2d60c..e489993 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -109,6 +109,33 @@
 	regmap_writel((u32)(offset & ~SDRAM_BASE), mmPREBOOT_VER_OFFSET);
 }
 
+#define FUSE_VER_PREFIX		"Fuse version "
+#define FUSE_VER_UNKNOWN	"unknown"
+static char preboot_fuse_ver_str[VERSION_MAX_LEN] = FUSE_VER_PREFIX;
+
+__weak int hl_fuse_get_devid(char *devid)
+{
+	return 1;
+}
+
+static void fuse_version(void)
+{
+	unsigned long offset = (unsigned long)preboot_fuse_ver_str;
+	char *devid = preboot_fuse_ver_str + sizeof(FUSE_VER_PREFIX) - 1;
+	const char *unknown = FUSE_VER_UNKNOWN;
+	int err, i;
+
+	err = hl_fuse_get_devid(devid);
+	if (err)
+		for (i = 0; i < sizeof(FUSE_VER_UNKNOWN); i++)
+			devid[i] = unknown[i];
+
+	preboot_puts_raw(preboot_fuse_ver_str);
+	preboot_puts("\n");
+
+	regmap_writel((u32)(offset & ~SDRAM_BASE), mmFUSE_VER_OFFSET);
+}
+
 __weak void preboot_security_init(void) {}
 __weak void preboot_led_on(void) {}
 
@@ -131,6 +158,7 @@
 
 	hl_ts_init();
 	preboot_version();
+	fuse_version();
 	preboot_led_on();
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
